src/Microsoft.AspNet.Mvc.DataAnnotations/DataAnnotationsModelValidator.cs

```diff
@@ -26,6 +27,115 @@ public DataAnnotationsModelValidator(ValidationAttribute attribute, IStringLocal
+                    {
+                        return new MinLengthAttributeAdapter((MinLengthAttribute)Attribute, _stringLocalizer)
+                            .GetErrorMessage(modelMetadata);
+                    }
+                },
+                {
+                    typeof(RegularExpressionAttribute), (modelMetadata) =>
+                    {
+                        return new RegularExpressionAttributeAdapter((RegularExpressionAttribute)Attribute, _stringLocalizer)
+                            .GetErrorMessage(modelMetadata);
+                    }
+                },
+                {
+                    typeof(CompareAttribute), (modelMetadata) =>
+                    {
+                        return new CompareAttributeAdapter((CompareAttribute)Attribute, _stringLocalizer)
+                            .GetErrorMessage(modelMetadata, metadataProvider);
+
+                    }
+                },
+                {
+                    typeof(StringLengthAttribute), (modelMetadata) =>
+                    {
+                        return new StringLengthAttributeAdapter((StringLengthAttribute)Attribute, _stringLocalizer)
+                            .GetErrorMessage(modelMetadata);
+                    }
+                },
+                {
+                    typeof(CreditCardAttribute), (modelMetadata) =>
+                    {
+                        return new DataTypeAttributeAdapter((DataTypeAttribute)Attribute, "creditcard", _stringLocalizer)
+                            .GetErrorMessage(modelMetadata);
+                    }
+                },
+                {
+                    typeof(EmailAddressAttribute), (modelMetadata) =>
+                    {
+                        return new DataTypeAttributeAdapter((DataTypeAttribute)Attribute, "email", _stringLocalizer)
+                            .GetErrorMessage(modelMetadata);
+                    }
+                },
+                {
+                    typeof(PhoneAttribute), (modelMetadata) =>
+                    {
+                        return new DataTypeAttributeAdapter((DataTypeAttribute)Attribute, "phone", _stringLocalizer)
+                            .GetErrorMessage(modelMetadata);
+                    }
+                },
+                {
+                    typeof(UrlAttribute), (modelMetadata) =>
+                    {
+                        return new DataTypeAttributeAdapter((DataTypeAttribute)Attribute, "url", _stringLocalizer)
+                            .GetErrorMessage(modelMetadata);
+                    }
+                }
+            };
+        }
+
+        private string GetErrorMessage(ModelMetadata metadata, IModelMetadataProvider metadataProvider)
+        {
+            if (Attribute == null)
+            {
+                throw new ArgumentNullException(nameof(Attribute));
+            }
+
+
+            var attrType = Attribute.GetType();
+
+            var dict = BuildErrorMessageDict(metadataProvider);
```

Comment:
The dictionary uses Attribute and _stringLocalizer, both of which are instance items, but it's probably worth it to make the function signature a little more gross by passing them there to avoid having to new that dictionary every time.